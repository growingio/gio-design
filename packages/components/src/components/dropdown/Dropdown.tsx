import React, { cloneElement } from 'react';
import { isFunction, isUndefined } from 'lodash';
import Tooltip from '../tooltip';
import { DropdownProps } from './interface';
import useControlledState from '../../utils/hooks/useControlledState';

const Dropdown = (props: DropdownProps) => {
  const placementList = ['top', 'bottom', 'topLeft', 'topRight', 'bottomLeft', 'bottomRight'];
  const {
    children,
    prefixCls,
    subPrefixCls = 'dropdown',
    placement = 'bottom',
    trigger = 'click',
    visible,
    onVisibleChange,
    overlay,
    ...rest
  } = props;
  const [controlledVisible, setControlledVisible] = useControlledState(visible, false);

  const getOverlay = () => {
    const _overlay: React.ReactElement = isFunction(overlay) ? overlay() : overlay;
    const onOverlayClick = (e: React.MouseEvent) => {
      setControlledVisible(false, true);
      onVisibleChange?.(false);
      _overlay.props.onClick?.(e);
    };
    return isUndefined(visible) ? cloneElement(_overlay, { onClick: onOverlayClick }) : _overlay;
  };

  return (
    <Tooltip
      prefixCls={prefixCls}
      subPrefixCls={subPrefixCls}
      trigger={trigger}
      placement={placementList.includes(placement) ? placement : 'bottom'}
      visible={controlledVisible}
      overlay={getOverlay()}
      onVisibleChange={(_visible) => {
        setControlledVisible(_visible);
        onVisibleChange?.(_visible);
      }}
      {...rest}
    >
      {children}
    </Tooltip>
  );
};

export default Dropdown;
