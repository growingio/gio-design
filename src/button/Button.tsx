import classNames from 'classnames';
import React from 'react';
import { LoadingOutlined } from '@gio-design/icons';
import { ButtonProps } from './interface';
import usePrefixCls from '../utils/hooks/use-prefix-cls';

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const {
    type = 'primary',
    size = 'normal',
    loading = false,
    disabled: disabledProp = false,
    htmlType = 'button',
    prefix,
    suffix,
    className,
    children,
    ...restProps
  } = props;

  const prefixCls = usePrefixCls('button');
  const classes = classNames([className, prefixCls], {
    [`${prefixCls}_${type}`]: type,
    [`${prefixCls}_${size}`]: size,
    [`${prefixCls}_loading`]: loading,
  });

  const prefixIcon = loading ? (
    <span className={`${prefixCls}-prefix-icon`}>
      <LoadingOutlined rotating />
    </span>
  ) : (
    prefix && <span className={`${prefixCls}-prefix-icon`}>{prefix}</span>
  );

  const suffixIcon = suffix && <span className={`${prefixCls}-suffix-icon`}>{suffix}</span>;

  const disabled = disabledProp || loading;

  const other: typeof restProps = {
    role: 'button',
    'aria-disabled': disabled,
  };

  return (
    <button
      ref={ref}
      // eslint-disable-next-line react/button-has-type
      type={htmlType}
      className={classes}
      disabled={disabled}
      data-testid="button"
      {...other}
      {...restProps}
    >
      {prefixIcon}
      {children}
      {suffixIcon}
    </button>
  );
});

Button.displayName = 'Button';

Button.defaultProps = {
  type: 'primary',
  size: 'normal',
  loading: false,
  disabled: false,
  htmlType: 'button',
};

export default Button;
