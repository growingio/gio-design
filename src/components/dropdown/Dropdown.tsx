import React, { cloneElement, Children } from 'react';
import { isFunction, isUndefined } from 'lodash';
import classnames from 'classnames';
import Tooltip from '../tooltip';
import { DropdownProps } from './interface';
import useControlledState from '../../utils/hooks/useControlledState';

const Dropdown = (props: DropdownProps) => {
  const placementList = ['top', 'bottom', 'topLeft', 'topRight', 'bottomLeft', 'bottomRight'];
  const {
    children,
    prefixCls,
    subPrefixCls = 'dropdown',
    placement = 'bottom',
    trigger = 'click',
    visible,
    onVisibleChange,
    overlay,
    ...rest
  } = props;
  const [controlledVisible, setControlledVisible] = useControlledState(visible, false);

  const getDropdownTrigger = () => {
    try {
      const child = Children.only(children);
      return cloneElement(child, {
        className: classnames(controlledVisible ? 'dropdown-active' : '', child.props.className),
      });
    } catch (e) {
      return children;
    }
  };

  const getOverlay = () => {
    const _overlay: React.ReactElement = isFunction(overlay) ? overlay() : overlay;
    const onOverlayClick = (e: React.MouseEvent) => {
      setControlledVisible(false, true);
      onVisibleChange?.(false);
      _overlay.props.onClick?.(e);
    };
    return isUndefined(visible) ? cloneElement(_overlay, { onClick: onOverlayClick }) : _overlay;
  };

  return (
    <Tooltip
      prefixCls={prefixCls}
      subPrefixCls={subPrefixCls}
      trigger={trigger}
      placement={placementList.includes(placement) ? placement : 'bottom'}
      visible={controlledVisible}
      overlay={getOverlay()}
      onVisibleChange={(_visible) => {
        setControlledVisible(_visible);
        onVisibleChange?.(_visible);
      }}
      {...rest}
    >
      {getDropdownTrigger()}
    </Tooltip>
  );
};

export default Dropdown;
