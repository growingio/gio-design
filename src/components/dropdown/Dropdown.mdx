import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Dropdown from './index';

# Dropdown 下拉菜单

<Subtitle>向下弹出的列表。当页面上的操作命令过多时，用此组件可以收纳操作元素。点击或移入触点，会出现一个下拉菜单。可在列表中进行选择，并执行相应的命令。</Subtitle>

[1] 用于收罗一组命令操作。

[2] Select 用于选择，而 Dropdown 是命令集合。


## 应用场景

【在控件中使用】

- Table - 「···」、「头像列表 - 看板」、「Tag 列表 - 埋点事件」、「筛选」
- 导航 - 「头像」、「选择项目」
- 搜索框 - 「最近搜索记录」

【在页面中使用】

- Button「···」触发；
- Select 触发。

## 参数说明

<ArgsTable of={Dropdown} />

## 代码演示

[1] 当页面上的操作命令过多时，常使用 Dropdown 收纳操作元素。

[2] 点击触点，出现下拉菜单，在列表中选择相应的命令执行，点击相应命令后自动收起列表，点击区域为整条列表项，点击空白处列表收起。

### 基本

[Example](/?path=/story/functional-components-dropdown--default)

### 位置

一共 12 种位置。位置选择原则：下拉菜单需要显示完整且不错乱。

[Example](/?path=/story/functional-components-dropdown--placement)
