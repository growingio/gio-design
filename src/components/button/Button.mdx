import { Subtitle, ArgsTable, Story, Canvas } from '@storybook/addon-docs/blocks';
import Button from '.';

# Button 按钮

<Subtitle>点击后即执行相应操作</Subtitle>

## 使用规范

主要按钮 primary

> 突出显示页面上最强烈的动作，应该只在每个屏幕上出现一次(不包括应用程序头或模态对话框)。并不是每个屏幕都需要主按钮。

次要按钮 secondary

> 用于与主按钮配对，以明确有两个选项；或者用于多个无主次之分的操作时。

辅助按钮 assist

> 当页面有许多高频率使用的操作时，将其用作默认按钮。如：看板页面。

文本按钮 text

> 适用于列表内添加多个内容时，如「事件分析」中「添加事件」、「fliter」中「添加过滤条件」等。

图标按钮 icon

> 图标 icon 的颜色可以自定义，图标按钮的存在两种样式：背景为透明度 0%的、浅蓝色背景的。

可扩展

> 按钮具有多种 padding。大多数情况下，我们使用默认紧凑型的 padding，即下列「间距」的规范。还可以将按钮扩展到全宽以填充父容器。

## 参数说明

<ArgsTable of={Button} />

## 代码演示

### 基础

<Canvas>
  <Story id="basic-components-button--default" />
</Canvas>

### 图标

<Canvas>
  <Story id="basic-components-button--icon-button" />
</Canvas>

### 可扩展

<Canvas>
  <Story id="basic-components-button--block-button" />
</Canvas>
