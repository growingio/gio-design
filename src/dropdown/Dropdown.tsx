import React, { cloneElement, Children, forwardRef } from 'react';
import classnames from 'classnames';
import { usePrefixCls } from '@gio-design/utils';
import { get, isFunction, isUndefined, noop } from 'lodash';
import DropdownProps from './interface';
import useControlledState from '../utils/hooks/useControlledState';
import Popover from '../popover';
import { BUTTON_DISPLAY_NAME } from '../button/Button';
import { ICON_BUTTON_DISPLAY_NAME } from '../button/IconButton';

export const Dropdown = forwardRef<HTMLDivElement, DropdownProps>((props, ref) => {
  const {
    children,
    placement = 'bottomLeft',
    trigger = 'click',
    defaultVisible = false,
    visible,
    onVisibleChange,
    content,
    overlayClassName,
    disabled,
    ...rest
  } = props;

  const [controlledVisible, setControlledVisible] = useControlledState(visible, defaultVisible);
  const prefixCls = usePrefixCls('dropdown');

  const getDropdownTrigger = () => {
    const child = Children.only(children);
    const childProps: {
      className?: string;
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      onClick?: (...args: any) => void;
      active?: boolean;
    } = {
      className: classnames(
        {
          'dropdown-active': controlledVisible,
        },
        child.props.className
      ),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      onClick: (...args: any) => {
        setControlledVisible(!controlledVisible);
        child.props.onClick?.(...args);
      },
    };
    if ([BUTTON_DISPLAY_NAME, ICON_BUTTON_DISPLAY_NAME].includes(get(child.type, 'displayName'))) {
      childProps.active = child.props.active ?? controlledVisible;
    }
    return cloneElement(child, childProps);
  };

  const withoutPadding = (contentNode: DropdownProps['content']) => {
    if (React.isValidElement(contentNode)) {
      return React.cloneElement(contentNode, {
        className: classnames(`${prefixCls}-p-0`, contentNode.props.className),
      });
    }
    return contentNode;
  };

  const withOnClick = (contentNode: DropdownProps['content']) => {
    const close = () => {
      setControlledVisible(false);
      if (isFunction(onVisibleChange)) {
        onVisibleChange(false);
      }
    };
    const onClick = close;

    return (
      <div role="none" ref={ref} onKeyDown={noop} onClick={isUndefined(visible) ? onClick : noop}>
        {contentNode}
      </div>
    );
  };

  return (
    <Popover
      disabled={disabled}
      trigger={trigger}
      arrowPointAtCenter
      placement={placement}
      visible={controlledVisible}
      content={withOnClick(withoutPadding(content))}
      overlayClassName={classnames(prefixCls, overlayClassName)}
      onVisibleChange={(changedVisible) => {
        setControlledVisible(changedVisible);
        if (isFunction(onVisibleChange)) {
          onVisibleChange(changedVisible);
        }
      }}
      data-testid="dropdown"
      {...rest}
    >
      {getDropdownTrigger()}
    </Popover>
  );
});

export default Dropdown;
