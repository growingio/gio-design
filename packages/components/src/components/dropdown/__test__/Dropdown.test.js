import React from 'react';
import Dropdown from '../index';
import Button from '../../button';
import { act } from 'react-dom/test-utils';
import { mount, render } from 'enzyme';

async function waitForComponentToPaint(wrapper, amount = 500) {
  await act(async () => new Promise((resolve) => setTimeout(resolve, amount)).then(() => wrapper.update()));
}

describe('Testing dropdown', () => {
  const getDropdown = () => (
    <Dropdown overlay="Dropdown 内容主体">
      <Button>Test</Button>
    </Dropdown>
  );

  it('should be stable', () => {
    const wrapper = render(getDropdown());
    expect(wrapper).toMatchSnapshot();
  });

  it('should be mount, setProps, unmount with no error', () => {
    expect(() => {
      const wrapper = mount(getDropdown());
      wrapper.setProps({ overlay: 'overlay update' });
      wrapper.setProps({ trigger: 'hover' });
      wrapper.setProps({ visible: 'true' });
      wrapper.unmount();
    }).not.toThrow();
  });

  it('should be render rightly', (done) => {
    const wrapper = mount(getDropdown());
    wrapper.setProps({ trigger: 'click' });
    wrapper.setProps({ placement: 'topLeft' });
    wrapper.find('button').at(0).simulate('click');
    expect(wrapper.exists('.gio-dropdown-inner')).toBe(true);
    waitForComponentToPaint(wrapper).then(() => {
      expect(wrapper.exists('.gio-dropdown-placement-topLeft')).toBe(true);
      done();
    });
  });
});
