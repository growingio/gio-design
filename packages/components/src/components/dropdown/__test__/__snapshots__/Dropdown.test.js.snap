// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing dropdown should be stable 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "gio-btn",
      "style": "cursor:pointer",
      "type": "button",
    },
    "children": Array [
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "data": "Test",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "style": undefined,
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "style": undefined,
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
